### PR TITLE
[Vectorized][Feature] add ColumnHLL to support hll type

### LINKED ISSUE #7826: [Vectorized][Feature]add ColumnHLL to support HLL type
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Description

_No response_

### Use case

_No response_

### Related issues

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by zhangstar333**
# Proposed changes

Issue Number: close #7826 

## Problem Summary:

Describe the overview of changes.

## Checklist(Required)

1. Does it affect the original behavior: (Yes/No/I Don't know)
2. Has unit tests been added: (Yes/No/No Need)
3. Has document been added or modified: (Yes/No/No Need)
4. Does it need to update dependencies: (Yes/No)
5. Are there any changes that cannot be rolled back: (Yes/No)

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

File: be/src/vec/data_types/data_type_hll.cpp [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+
+#include "vec/data_types/data_type_hll.h"
+
+#include "vec/columns/column_complex.h"
+#include "vec/common/assert_cast.h"
+#include "vec/io/io_helper.h"
+
+namespace doris::vectorized {
+
+size_t DataTypeHLL::serialize(const IColumn& column, PColumn* pcolumn) const {
```
Comment (morningman) [inline be/src/vec/data_types/data_type_hll.cpp:26]: The serialize() method has been refactored in #7939 , please change it after that PR being merged.

File: be/src/vec/data_types/data_type_bitmap.h [hunk 2]
```diff
@@ -83,6 +84,9 @@ struct is_complex : std::false_type {};
 template <>
 struct is_complex<DataTypeBitMap::FieldType> : std::true_type {};
 
+template <>
+struct is_complex<DataTypeHLL::FieldType> : std::true_type {};
```
Comment (morningman) [inline be/src/vec/data_types/data_type_bitmap.h:88]: Why define HLL in `data_type_bitmap`?

File: be/src/vec/data_types/data_type_hll.cpp (hunk 1 above)
Comment (zhangstar333) [reply be/src/vec/data_types/data_type_hll.cpp:26]: OK!

File: be/src/vec/aggregate_functions/aggregate_function_hll_union_agg.h [hunk 3]
```diff
@@ -116,12 +114,11 @@ class AggregateFunctionHLLUnion final : public AggregateFunctionHLLUnionAgg {
...
 
-    DataTypePtr get_return_type() const override { return std::make_shared<DataTypeString>(); }
+    DataTypePtr get_return_type() const override { return std::make_shared<DataTypeHLL>(); } 
 
     void insert_result_into(ConstAggregateDataPtr __restrict place, IColumn& to) const override {
-        auto& column = static_cast<ColumnString&>(to);
-        auto result = this->data(place).get();
-        column.insert_data(result.c_str(), result.length());
+        auto& column = static_cast<ColumnHLL&>(to);
+        column.get_data().push_back(this->data(place).get());
```
Comment (HappenLee) [inline be/src/vec/aggregate_functions/aggregate_function_hll_union_agg.h:121]: emplace_back to reduce unless copy

File: be/src/vec/data_types/data_type_bitmap.h (hunk 2 above)
Comment (zhangstar333) [reply be/src/vec/data_types/data_type_bitmap.h:88]: > Why define HLL in `data_type_bitmap`?

I will deal with it

File: be/src/vec/functions/hll_empty.cpp [hunk 4]
```diff
@@ -20,19 +20,21 @@
...
 struct HLLEmptyImpl {
     static constexpr auto name = "hll_empty";
-    static auto get_return_type() { return std::make_shared<DataTypeString>(); }
-    static Field init_value() {
+    using ReturnColVec = ColumnHLL;
+    static auto get_return_type() { return std::make_shared<DataTypeHLL>(); }
+    static HyperLogLog init_value() {
         auto hll = HyperLogLog::empty();
-        return {hll.c_str(), hll.size()};
+        return HyperLogLog(hll);
```
Comment (HappenLee) [inline be/src/vec/functions/hll_empty.cpp:33]: directly construct the class

File: be/src/vec/functions/hll_hash.cpp [hunk 5]
```diff
@@ -20,73 +20,47 @@
...
-
-        size_t prev_offset = 0;
-        size_t res_offset = 0;
 
         for (size_t i = 0; i < size; ++i) {
-            std::string hash_string;
             if (nullmap[i]) {
-                hash_string = HyperLogLog::empty();
+                std::string hash_string = HyperLogLog::empty();
+                res_data[i] = HyperLogLog(Slice(hash_string));
```
Comment (HappenLee) [inline be/src/vec/functions/hll_hash.cpp:57]: here should direct continue

**Review by HappenLee — COMMENTED** [review]
some problem need to fix

File: be/src/vec/core/block.cpp [hunk 6]
```diff
@@ -810,6 +814,14 @@ doris::Tuple* Block::deep_copy_tuple(const doris::TupleDescriptor& desc, MemPool
             string_slot->ptr = reinterpret_cast<char*>(pool->allocate(size));
             bitmap_value->write(string_slot->ptr);
             string_slot->len = size;
+        } else if (slot_desc->type() == TYPE_HLL) {
+            auto hll_value = (HyperLogLog*)(data_ref.data);
+            LOG(INFO)<<"block hll string: "<<hll_value->to_string();
```
Comment (HappenLee) [inline be/src/vec/core/block.cpp:819]: del the log(INFO)

File: be/src/vec/core/block.cpp [hunk 7]
```diff
@@ -810,6 +814,14 @@ doris::Tuple* Block::deep_copy_tuple(const doris::TupleDescriptor& desc, MemPool
...
+            auto size = hll_value->max_serialized_size();
+            auto string_slot = dst->get_string_slot(slot_desc->tuple_offset());
+            string_slot->ptr = reinterpret_cast<char*>(pool->allocate(size));
+            hll_value->serialize((uint8_t*)string_slot->ptr);
+            string_slot->len = size;
```
Comment (HappenLee) [inline be/src/vec/core/block.cpp:824]: set `string_slot->len` to serialize size

File: be/src/vec/data_types/data_type_hll.cpp [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+        hll_size_array[i + 1] = hll.max_serialized_size();
+    }
+
+    auto allocate_len_size = sizeof(size_t) * (column_num + 1);
+    memcpy(buf, hll_size_array, allocate_len_size);
+    buf += allocate_len_size;
+
+    for (size_t i = 0; i < column_num; ++i) {
+        auto& hll = const_cast<HyperLogLog&>(data_column.get_element(i));
+        hll.serialize(reinterpret_cast<uint8_t*>(buf));
```
Comment (HappenLee) [inline be/src/vec/data_types/data_type_hll.cpp:45]: the size is not match to `max serialized size` may cause error

**Review by HappenLee — COMMENTED** [review]
some little problem

File: be/src/vec/data_types/data_type_hll.cpp [hunk 9]
```diff
@@ -76,10 +79,12 @@ int64_t DataTypeHLL::get_uncompressed_serialized_bytes(const IColumn& column) co
     auto allocate_content_size = 0;
     for (size_t i = 0; i < column.size(); ++i) {
         auto& hll = const_cast<HyperLogLog&>(data_column.get_element(i));
-        allocate_content_size += hll.max_serialized_size();
+        std::string result(hll.max_serialized_size(), '0');
+        size_t actual_size = hll.serialize((uint8_t*)result.c_str());
```
Comment (HappenLee) [inline be/src/vec/data_types/data_type_hll.cpp:83]: should not serialize two times

**Review by HappenLee — APPROVED** [review]
LGTM

**Pull request merged**